### ISSUE TITLE
[39] Language Modelling at Scale: Gopher, Ethical considerations, and Retrieval

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#67]
[blog](https://deepmind.com/research/publications/2021/scaling-language-models-methods-analysis-insights-from-training-gopher)
[paper](https://kstatic.googleusercontent.com/files/b068c6c0e64d6f933068f7de30ea722359ef87c6c14d3065856b86d44fbdf2dea3ff373ed9eb751514f242d20df9d6a468622fad093f962563545e7d0cdb9dba)

280B model gopher 관련 논문
- section 3
    - Curation of MassiveText
    - Architecture
    - Optimization
    - Infrastructure
- section 4
    - 152 개의 set에서 benchmark
- section 5
    - Examine model toxity and bias
- section 6
    - Dialogue interaction
- section 7 
    - Ethical and Safe applications

# 3. Method

## 3.1 Models
GPT-2 모델 그대로 가져다 쓰는데, 2가지가 바뀐다.
1. LayerNorm => RMSNorm
2. Absolute PE => Relative PE (Longer sequence 처리 위함)

tokenizer 는 SentencePiece 사용 
(32,000 vocab, byte-level back off)
(open vocabulary modelling 위함)

![image](https://user-images.githubusercontent.com/16400591/145543279-39c0e723-6982-471f-a785-ca387693d177.png)

## 3.2 Training
모든 모델을 다음으로 학습시켰다.
1. 300 billion token
2. 2048 token context window 
3. Adam 
    1. lr 10e-7 부터 1500 step 동안 warmup
    2. 4e-5 lr
    3. 3e-5 lr 까지 cosine 으로 decay
    4. clip gradient norm (`1.0`, 7.1B 모델에서는 `0.25`)
4. batch schedule
    1. 3 million => 6 million
5. bfloat16 (with fp32 optimizer)

## 3.3 InfraStructure
1. JAX & Haiku code base
2. JAX의 pmap 으로 data, model parallel 달성
3. TPUv3
4. half-precision parameter 에 single-precision Adam 은 2.5TiB 정도 먹었음
5. optimizer state partitioning 사용
6. model parallelism 사용
7. rematerialization 사용
8. pipelining 은 미사용 (TPU 에서는 별 효과 없음)

## 3.4 Training Dataset
MassiveText curation
- 영문 dataset
- 2.35 B documents
- 10.5 TB Text
- gopher 가 300 B token 으로 학습되서 sub-sample 은 필수. 요 녀석도 tuning 하면 좋음 (appendix A.3.1) 

![image](https://user-images.githubusercontent.com/16400591/145704672-37dabbfd-8acb-4d7f-8c2e-de284fff5585.png)

Dataset 구할 때 사용한 pipeline 은 appendix A.1.1 참조
여러가지 테크닉이 downstream 성능에까지 좋은 영향 끼침 appendix A.3.2

# 4. Results
152 task 중 124 개 task 에서 SOTA 찍음
![image](https://user-images.githubusercontent.com/16400591/145705018-5ab02c44-6960-4466-afc4-491642bc2b72.png)

아래는 전체적인 결과를 현 SOTA 모델 기준으로 얼마나 성능이 올랐는지, 떨어졌는지를 시각화 한 그림이다.
해당 그림에는 `GPT-3(175B)`, `Jurassic-1(178B)`, `Megatron-Turing NLG(530B)` 결과를 포함한다.
`gopher 280B` 모델은 상당히 좋았다 !
![image](https://user-images.githubusercontent.com/16400591/145705064-816780ad-6cf3-42f6-99fc-2421f6ee1fda.png)

gopher 가 강한 task
- reading comprehension
- humanities
- ethics
- STEM
- medicine categories

적절히 잘하는 task
- fact-checking

아직 어려워 하는 task
- common sense reasoning
- logical reasoning
- math

# 5. Toxicity and Bias Analysis
모델이 얼마나 toxic 한 결과를 내뱉는 지 분석.
여러 논문들이 현재 evaluation을 지적함. 그에 따라 좀 더 많은 것을 분석했음 (section 7.2)
물론 현재 eval metric 수치도 뽑았음

[Perspective API](https://perspectiveapi.com) 로 toxicity 수치를 구했음
더 이상은 생략.

# 6 Dialogue
finetuning 굳이 안하고 prompt 제공하면 좋은 성능 나옴.
![image](https://user-images.githubusercontent.com/16400591/145705730-2c7f29b6-c46e-4ec9-a56e-76d25ff817da.png)

위 예제를 보면, 독백이나 블로그 포스트가 나오는데, 학습한 데이터를 생각하면 정상.
성공 케이스와 실패 케이스
![image](https://user-images.githubusercontent.com/16400591/145705901-7432491e-9e36-4d82-a045-922cf7faba58.png)

DialoguePrompted Gopher는 여전히 LM일 뿐임을 강조함. prompt는 모델의 response를 조건화하지만 일관되게 신뢰할 수 있거나 사실적인 대화를 하도록 만드는 게 아님.

# Appendix A  MassiveText
### A.1.1 Pipeline stages
![image](https://user-images.githubusercontent.com/16400591/145706382-be21ddca-e519-42d0-ba0a-f5568f4f3c29.png)

1. content filtering (All subsets)
    1. non-English 제거
    2. Google SafeSearch filter 통과 못하는 애들 제거 (MassiveWeb only)
2. Text Extraction (MassiveWeb only)
    1. HTML markup 구조로 데이터 추출
    2. high quality website 에서 보통 tree depth 같은 종류의 content 가 발견되는 점 이용
3. Quality Filtering (MassiveWeb only)
    1. 50~100,000 단어만 허용
    2. 평균 단어 길이 3~10 만 허용
    3. symbol-to-word ratio 가 0.1 보다 크다면 제거
    4. 90% 이상의 줄이 bullet 으로 시작하면 제거
    5. 30% 이상의 줄이 생략기호로 끝나도 제거
    6. 문서 기준 80% 이상의 단어가 영단어 포함해야 함
    7. stop-word filter 적용. (the, be, to, of, and, that, have, with) 가 포함되어 있지 않으면 제거
4. Repetition Removal (MassiveWeb only)
5. Document Deduplication (All subsets)
6. Test-set Filtering (All subsets)

### A.1.2. Constructing Token Sequences
1. MassiveText subset 에서 B byte 크기의 document 를 uniform 확률로 고름 
2. `C =15 × n` 만큼 crop. n은 token sequence length. crop 의 start-index s는 [-c/4, B-c/4) 사이에서 고름. [max(0, s), min(B, s + c)] crop 수행
3. tokenize + bos, eos 붙이기
4. 대부분 document 가 모델 seq_len 2048 보다 작아서 10 개의 tokenized를 concat 했음
5. (4) output 을 2048로 split 시킴 (i.e. (2048, 2048, 1900) )  마지막 chunk 는 버림
6. sampling
7. shuffle and batch training

## A.2. Dataset Analysis
생략

### A.3.1. MassiveText Subsets Weighting
![image](https://user-images.githubusercontent.com/16400591/145706557-a3e0d768-e921-4477-b684-89d3cdef0536.png)
subset 에서 batch 구성할 때 sampling weight를 어떻게 주느냐에 따라 성능 천차만별.

최종적으로 사용한 sampling ratio
- 10% C4
- 50% MassiveWeb
- 30% Books
- 10% News performs

### A.3.2. Iterative Refinement of MassiveWeb
![image](https://user-images.githubusercontent.com/16400591/145990840-f8cc662d-46d7-493b-8c55-bb246498c29d.png)

### A.4. Text normalization
tokenizer에서 NKFC16 text normalization 수행 (https://unicode.org/reports/tr15/)
손실은 발생함. `2^5` 는 `2 5` 로 바뀜.
future work 에서는 normalization 없이 open-vocab 에서 잘 수행되는 모델을 찾아낼 것임.

### A.5. MassiveText Datasheet
생략

# Appendix B.  Gopher Model Card
생략

# Appendix C  Lesson Learned
## C.1 Adafactor
memory footprint 를 줄이려고 adafactor를 써봤는데 성능 떨어짐.
작은 규모에서는 괜찮았음. (1.4B 까지는 ㄱㅊ)
LR 을 줄여도 7.1B 모델에서는 불안정했음
![image](https://user-images.githubusercontent.com/16400591/145704103-40c78c2b-6eff-4757-9f67-cf60a7779240.png)

## C.2 Lower-Precision Training with bfloat16
float16 만 쓰면 당연히 안 좋음.
417M 모델에서 ablation study를 좀 해봤는데,
bfloat16을 쓰려면 최소한 optimizer state는 fp32를 유지해야함.
![image](https://user-images.githubusercontent.com/16400591/145704157-cc1d0c7f-5a33-4894-aad7-cad41a97233c.png)

- `FP32 EveryWhere` : 말 그대로 param, activation 모두 float32
- `bfloat16 Parameters without Random Rounding` : param, activation 모두 bfloat16
- `bfloat16 Parameters with Random Rounding` : param, activation 모두 bfloat16, 근데 모델 forward 시킬 때, parameter가 랜덤하게 round-up 되거나 down 됨 (bfloat16 표현에 맞춰서 실제 value 값을 비교해 보면서 랜덤 prob을 조정함. 예를 들어, bfloat 이 0 과 1을 나타낼 수 있는데, 실제 값이 0.7 이면 `ceil()` 될 확률이 높음)
- `bloat16 parameters with a float32 copy in the partitioned optimiser state` : param, activation 모두 bfloat16, parameter-copy 본이 float32 로 optimizer-state에 저장되고 update 에 활용됨. forward path에는 random 하게 bfloat16으로 바뀜.

fp32 만 안 쓰면 1.4배정도 속도 향상이 있었음

# Appendix D.  Results
생략.

# Appendix E.  Toxicity and Bias Analysis
생략.

# Appendix F.  Compute Usage
![image](https://user-images.githubusercontent.com/16400591/145706108-8125aae5-c504-4fe1-a19c-36ad153988e1.png)
기존과 다르게 flop 에 reparametrization trick, padding, repeated computation 같은 것들까지 넣어서 계산했음.
repeat computation 을 빼서 계산하면 4~100배가량 속도가 빨라지는 것처럼 보임. 그렇게 리포트 하면 의미가 없음.

모델이 커질 때, 신기한 점은,
1. attention 에서 소모되는 시간 비율이 점점 줄어듦
2. attention computation 기준으로는 70% 이상의 computation 이 relative pos-emb 에서 이루어짐.
![image](https://user-images.githubusercontent.com/16400591/145706170-35e9a372-5f25-444e-a74d-f39fc98f3aec.png)

# Appendix G. Reducing Inference and Training Costs
## G.1. Efficient Fine-tuning
![image](https://user-images.githubusercontent.com/16400591/146003363-fec0c5be-2cd6-4e27-a034-4826da8a4579.png)

3가지 옵션 고려
1. bias only training
2. 마지막 40% layer 만 train
3. full finetune

full finetune 에 lr 만 잘 주면 같은 budget에서 성능은 젤 좋음

### G.2.1. Distillation
distill로 압축해보고 싶었는데 목표였던 성능을 유지하면서 1/2로 줄이기는 달성할 수 없었음. 어려움.
효과는 있긴 있음.
feature distillation 은 안해봤고, output distillation 수행했음
![image](https://user-images.githubusercontent.com/16400591/146003857-e7786cfa-bbe0-4a2e-b973-d1ae0b945bb1.png)
![image](https://user-images.githubusercontent.com/16400591/146003923-29f42c7b-ca59-4cb5-aba3-99f5dfb4eb64.png)

### G.2.2. Pruning
pruning도 별로 쓸만하지 않은 것 같음. pruning 결과로 gpu 성능을 향상시키자면 파라미터 수를 많이 줄여야 하는데 그 상황에서 성능 유지가 까다로움. cpu에서는 좀 낫겠지만 이 모델을 cpu에서 돌릴 일이 있겠나.

unstructured pruning 도 rough 한 방법만 사용해 본 것은 아쉬움. structured pruning 과 유명한 technique들의 결과도 궁금함.
![image](https://user-images.githubusercontent.com/16400591/146004511-056d2b50-cf0a-4722-9c7f-5ce583703ee5.png)

### G.3.1. Dynamic Sparse Training
dynamic sparse training (rigging the lottery) 잘 안 먹힘

### G.3.2. Reverse Distillation
reverse distillation. small pretrained model로 큰 모델에 distillation해서 scratch-training 가속하기. 학습 초반에만 잠깐 효과 있어서 큰 의미가 없음.
![image](https://user-images.githubusercontent.com/16400591/146004226-632cd154-5cd9-4238-b89b-82bc0d0e36ce.png)

### G.3.3. Warm starting
warm starting. 작은 모델의 width나 depth를 키우기. pretrained-model 에 새롭게 initialize 시킨 width, depth 를 키워서 학습하는 방식.

기본적으로 depth를 늘릴 때는, 각 layer 의 weight를 그대로 가져왔음
예를 들어,
`A, B, C, D, E` layer 를 5depth 늘린다면
`A, A, B, B, C, C, D, D, E, E` weight로 init 하고 train.

width 를 늘릴 때,
attention head weight matrix 는 tiling으로 늘려줌
![image](https://user-images.githubusercontent.com/16400591/146007765-592b3a12-8ca0-4ccf-83a0-8a9ef81d264e.png)

꽤 효과적이긴 했는데, 큰 모델에서는 width 키우는 건 별로 안 좋았음
3B parameter 모델을 `depth 1.5x, width 1.25x` 해봤는데, 오히려 약간 결과가 안좋았음

![image](https://user-images.githubusercontent.com/16400591/146004418-2a0cdfb8-1aa1-4e0e-beb8-4f8c43426e47.png)
[1 row] Warm starting of a 4.7B model from a 1.3B model
[2 row] Warm starting of a 9B model from a 4.5B model

### G.3.4. Alternative Warm Starting Methods
다른 init 방법들을 써봤는데 다 구렸음 (scratch 보다 안좋게 나옴)
• freshly initializing all new weights;
• drawing from the weight distributions of existing weights – especially when adding width to our models;
• initializing the new weights to very small values s.t. the behaviours of the original model is nearly preserved.

## G.4. Future Work for Efficient Training
아래는 나중에 해봐도 좋을 듯
- 새로운 architecture 적용
- MoE
- quantization
- hardware accelerated sparsity
- semi-parameteric approach

# Appendix H.  Dialogue Prompted Gopher Details
생략
